### PR TITLE
Refactor and extend file format tests

### PR DESCRIPTION
**Pull request by phoerious**
## Description

This is a follow-up on #1393 to fix, improve and extend critical file format tests.

## Motivation and context

The previous tests for KDBX loading and saving were rather obscure, not designed for use with multiple different file formats and some also weren't executed at all.

This patch

1. consolidates the confusing mess of different files for similar things into one base class and three specialized classes for KDBX 2, 3.x and 4
2. adds additional checks inside existing tests as well as additional tests (mostly version checks and thorough matrix tests for KDBX 3 -> 4 upgrades)
3. removes obsolete ToDbExporter tests (class does not exist anymore, tests weren't compiled anyway)

The first two of the above-mentioned steps increase test coverage for KDBX 4 significantly and reduce the amount of work required to write tests for all KDBX versions by ensuring that all version-agnostic tests are run on both KDBX 3.x and 4. KDBX 2 has been excluded here, because we do not support writing KDBX 2 files (saving a KDBX 2 file automatically upgrades to KDBX 3.1, which I also added a test for).

## How has this been tested?

Well... by running the tests.

## Types of changes

- ✅ Bug fix (non-breaking change which fixes an issue)

## Checklist:

- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**
- ✅ I have added tests to cover my changes.

### CODE REVIEW

File: tests/TestKdbx3.cpp [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+    QScopedPointer<Database> db(reader.readDatabase(filename, key));
+    QCOMPARE(reader.version(), KeePass2::FILE_VERSION_3);
+    QVERIFY(db.data());
+    QVERIFY(!reader.hasError());
+
+    QCOMPARE(db->rootGroup()->name(), QString("Format300"));
+    QCOMPARE(db->metadata()->name(), QString("Test Database Format 0x00030000"));
+}
+
+void TestKdbx3::testNonAscii()
```
Comment (louib) [inline tests/TestKdbx3.cpp]: @phoerious is it only the password that is non ascii? in which case we should rename that `testNonAsciiPassword`

File: tests/TestKdbx4.cpp [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    QFETCH(Uuid, kdfUuid);
+    QFETCH(Uuid, cipherUuid);
+    QFETCH(quint32, expectedVersion);
+
+    QScopedPointer<Database> sourceDb(new Database());
+    sourceDb->metadata()->setName("Wubba lubba dub dub");
+    QCOMPARE(sourceDb->kdf()->uuid(), KeePass2::KDF_AES_KDBX3);    // default is legacy AES-KDF
+
+    CompositeKey key;
+    key.addKey(PasswordKey("I am in great pain, please help me!"));
```
Comment (louib) [inline tests/TestKdbx4.cpp]: :ambulance:

File: tests/TestKdbx4.cpp [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+    QTest::newRow("Argon2           + AES")      << KeePass2::KDF_ARGON2    << KeePass2::CIPHER_AES      << kdbx4;
+    QTest::newRow("AES-KDF          + AES")      << KeePass2::KDF_AES_KDBX4 << KeePass2::CIPHER_AES      << kdbx4;
+    QTest::newRow("AES-KDF (legacy) + AES")      << KeePass2::KDF_AES_KDBX3 << KeePass2::CIPHER_AES      << kdbx3;
+    QTest::newRow("Argon2           + ChaCha20") << KeePass2::KDF_ARGON2    << KeePass2::CIPHER_CHACHA20 << kdbx4;
+    QTest::newRow("AES-KDF          + ChaCha20") << KeePass2::KDF_AES_KDBX4 << KeePass2::CIPHER_CHACHA20 << kdbx4;
+    QTest::newRow("AES-KDF (legacy) + ChaCha20") << KeePass2::KDF_AES_KDBX3 << KeePass2::CIPHER_CHACHA20 << kdbx3;
+    QTest::newRow("Argon2           + Twofish")  << KeePass2::KDF_ARGON2    << KeePass2::CIPHER_TWOFISH  << kdbx4;
+    QTest::newRow("AES-KDF          + Twofish")  << KeePass2::KDF_AES_KDBX4 << KeePass2::CIPHER_TWOFISH  << kdbx4;
+    QTest::newRow("AES-KDF (legacy) + Twofish")  << KeePass2::KDF_AES_KDBX3 << KeePass2::CIPHER_TWOFISH  << kdbx3;
```
Comment (louib) [inline tests/TestKdbx4.cpp]: @phoerious I wasn't aware of QT's data driven testing capabilities... this is neat!

File: tests/TestKdbx4.cpp (hunk 2 above)
Comment (phoerious) [reply tests/TestKdbx4.cpp]: Hey, look, I turned myself into a pickle!

File: tests/TestKdbx4.cpp (hunk 3 above)
Comment (phoerious) [reply tests/TestKdbx4.cpp]: It's pretty amazing to work with.

File: tests/TestKdbx3.cpp (hunk 1 above)
Comment (phoerious) [reply tests/TestKdbx3.cpp]: I think it's just the password, but I'm not entirely sure. It's one of the tests I copied over.

**Pull request merged**